keyboards/kbd67/hotswap/keymaps/zunger/keymap.c

```diff
@@ -0,0 +1,163 @@
+   * `----------------------------------------------------------------'
+  */
+[_MAGIC] = LAYOUT(
+ KC_GRV,  H(00b9), H(00b2), H(00b3), H(2074), H(2075), H(2076), H(2077), H(2078), H(2079), H(2070), H(207b), H(207a), H(2295), H(2260), _______, \
+ _______, H(00b1), H(03c9), H(03b5), H(03c1), H(03c4), H(03b8), H(03c5), H(03b9), H(03bf), H(03c0), H(2203), H(2200), _______,          _______, \
+ _______, H(03b1), H(03c3), H(03b4), H(03c6), H(03b3), H(03b7), H(2205), H(03ba), H(03bb), H(2228), H(2227),          _______,          _______, \
+ _______, H(03b6), H(03be), H(03c7), H(03c8), H(03b2), H(03bd), H(03bc), H(2272), H(2273), H(2208),                   _______, H(2191), H(2194), \
+ _______, _______, _______,                   _______,                            _______, _______,                   H(2190), H(2193), H(2192)),
+
+  /* Mathematical upper layer. Blank keys are TRNS.
+   * ,----------------------------------------------------------------.
+   * | `  | ₁| ₂ | ₃ | ₄ | ₅ | ₆ | ₇ | ₈ | ₉ | ₀ | ₋ | ₊ | ⊗ | ≈ |    |
+   * |----------------------------------------------------------------|
+   * |     | ∓ | Ω | Ε | Ρ | Τ | Θ | Υ | Ι | Ο | Π | ∄ | ∞ |     |    |
+   * |----------------------------------------------------------------|
+   * |       | Α | Σ | Δ | Φ | Γ | Η | ¬ | Κ | Λ | ∪ | ∩ |       |    |
+   * |----------------------------------------------------------------|
+   * |        | Ζ | Ξ | Χ | Ψ | Β | Ν | Μ | ⊆ | ⊇ | ∉ |      | ⇑ | ⇔  |
+   * |----------------------------------------------------------------|
+   * |    |    |    |                       |   |    ||||| ⇐ | ⇓ | ⇒  |
+   * `----------------------------------------------------------------'
+  */
+[_SHIFTMAGIC] = LAYOUT(
+ KC_GRV,  H(2081), H(2082), H(2083), H(2084), H(2085), H(2086), H(2087), H(2088), H(2089), H(2080), H(208b), H(208a), H(2297), H(2248), _______, \
+ _______, H(2213), H(03a9), H(0395), H(03a1), H(03a4), H(0398), H(03a5), H(0399), H(039f), H(03a0), H(2204), H(221e), _______,          _______, \
+ _______, H(0391), H(03a3), H(0394), H(03a6), H(0393), H(0397), H(00ac), H(039a), H(039b), H(222a), H(2229),          _______,          _______, \
+ _______, H(0396), H(039e), H(03a7), H(03a8), H(0392), H(039d), H(039c), H(2286), H(2287), H(2209),                   _______, H(21d1), H(21d4), \
+ _______, _______, _______,                   _______,                            _______, _______,                   H(21d0), H(21d3), H(21d2)),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  // We want magic to move us to layer _MAGIC, and shift+magic to move us to layer _SHIFTMAGIC. We
+  // can't do this with MO's because they get confused about the various orders in which keys can be
+  // pressed and released, so let's do it carefully and in code instead.
+  static bool shifted = false;
+  static bool magic = false;
+
+  if (keycode == KC_LSHIFT || keycode == KC_RSHIFT) {
```

Comment:
You could actually run `get_mods` here, instead.  It may be a better/more reliable solution for what you want. 

But then again, this should work just fine, too.

Reply:
Alas, I tried it, and ran into a problem: real_mods is updated in process_action, which runs after process_record_user. So get_mods doesn't have action to the current mod state when this function runs.

That said, I did restructure this function to make a single call to layer_state_set, instead of doing the multiple layer changes, after reading what those functions did.

(I also added a SAFE_RESET to not trigger RESET on Fn+Esc after doing that by accident one time too many today. It's now Fn+LShft+Es.)

Reply:
Ah, okay. That makes sense then. (and I found another spot where post processing may be useful then)